File: paimon-core/src/main/java/org/apache/paimon/table/system/PartitionsTable.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.table.system;

import org.apache.paimon.CoreOptions;
import org.apache.paimon.casting.CastExecutor;
import org.apache.paimon.casting.CastExecutors;
import org.apache.paimon.catalog.Catalog;
import org.apache.paimon.catalog.CatalogLoader;
import org.apache.paimon.catalog.Identifier;
import org.apache.paimon.data.BinaryRow;
import org.apache.paimon.data.BinaryString;
import org.apache.paimon.data.GenericRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.data.Timestamp;
import org.apache.paimon.data.serializer.InternalRowSerializer;
import org.apache.paimon.disk.IOManager;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.manifest.PartitionEntry;
import org.apache.paimon.options.Options;
import org.apache.paimon.partition.Partition;
import org.apache.paimon.predicate.Predicate;
import org.apache.paimon.reader.RecordReader;
import org.apache.paimon.table.FileStoreTable;
import org.apache.paimon.table.ReadonlyTable;
import org.apache.paimon.table.Table;
import org.apache.paimon.table.source.InnerTableRead;
import org.apache.paimon.table.source.InnerTableScan;
import org.apache.paimon.table.source.ReadOnceTableScan;
import org.apache.paimon.table.source.SingletonSplit;
import org.apache.paimon.table.source.Split;
import org.apache.paimon.table.source.TableRead;
import org.apache.paimon.table.source.snapshot.TimeTravelUtil;
import org.apache.paimon.types.BigIntType;
import org.apache.paimon.types.DataField;
import org.apache.paimon.types.DataType;
import org.apache.paimon.types.DataTypes;
import org.apache.paimon.types.RowType;
import org.apache.paimon.utils.InternalRowPartitionComputer;
import org.apache.paimon.utils.InternalRowUtils;
import org.apache.paimon.utils.IteratorRecordReader;
import org.apache.paimon.utils.JsonSerdeUtil;
import org.apache.paimon.utils.ProjectedRow;
import org.apache.paimon.utils.SerializationUtils;

import org.apache.paimon.shade.guava30.com.google.common.collect.Iterators;

import java.io.IOException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import static org.apache.paimon.catalog.Identifier.SYSTEM_TABLE_SPLITTER;

/** A {@link Table} for showing partitions info. */
public class PartitionsTable implements ReadonlyTable {

    private static final long serialVersionUID = 1L;

    public static final String PARTITIONS = "partitions";

    public static final RowType TABLE_TYPE =
            new RowType(
                    Arrays.asList(
                            new DataField(0, "partition", SerializationUtils.newStringType(true)),
                            new DataField(1, "record_count", new BigIntType(false)),
                            new DataField(2, "file_size_in_bytes", new BigIntType(false)),
                            new DataField(3, "file_count", new BigIntType(false)),
                            new DataField(4, "last_update_time", DataTypes.TIMESTAMP_MILLIS()),
                            new DataField(5, "created_at", DataTypes.TIMESTAMP_MILLIS()),
                            new DataField(6, "created_by", DataTypes.STRING()),
                            new DataField(7, "updated_by", DataTypes.STRING()),
                            new DataField(8, "options", DataTypes.STRING())));

    private final FileStoreTable storeTable;

    public PartitionsTable(FileStoreTable storeTable) {
        this.storeTable = storeTable;
    }

    @Override
    public String name() {
        return storeTable.name() + SYSTEM_TABLE_SPLITTER + PARTITIONS;
    }

    @Override
    public RowType rowType() {
        return TABLE_TYPE;
    }

    @Override
    public List<String> primaryKeys() {
        return Collections.singletonList("partition");
    }

    @Override
    public FileIO fileIO() {
        return storeTable.fileIO();
    }

    @Override
    public InnerTableScan newScan() {
        return new PartitionsScan();
    }

    @Override
    public InnerTableRead newRead() {
        return new PartitionsRead(storeTable);
    }

    @Override
    public Table copy(Map<String, String> dynamicOptions) {
        return new PartitionsTable(storeTable.copy(dynamicOptions));
    }

    private static class PartitionsScan extends ReadOnceTableScan {

        @Override
        public InnerTableScan withFilter(Predicate predicate) {
            return this;
        }

        @Override
        public Plan innerPlan() {
            return () -> Collections.singletonList(new PartitionsSplit());
        }
    }

    private static class PartitionsSplit extends SingletonSplit {

        private static final long serialVersionUID = 1L;

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            return o != null && getClass() == o.getClass();
        }

        @Override
        public int hashCode() {
            return 1;
        }
    }

    private static class PartitionsRead implements InnerTableRead {

        private final FileStoreTable fileStoreTable;

        private RowType readType;

        public PartitionsRead(FileStoreTable table) {
            this.fileStoreTable = table;
        }

        @Override
        public InnerTableRead withFilter(Predicate predicate) {
            // TODO
            return this;
        }

        @Override
        public InnerTableRead withReadType(RowType readType) {
            this.readType = readType;
            return this;
        }

        @Override
        public TableRead withIOManager(IOManager ioManager) {
            return this;
        }

        @Override
        public RecordReader<InternalRow> createReader(Split split) throws IOException {
            if (!(split instanceof PartitionsSplit)) {
                throw new IllegalArgumentException("Unsupported split: " + split.getClass());
            }

            List<Partition> partitions = listPartitions();

            List<DataType> fieldTypes =
                    fileStoreTable.schema().logicalPartitionType().getFieldTypes();
            InternalRow.FieldGetter[] fieldGetters =
                    InternalRowUtils.createFieldGetters(fieldTypes);
            List<CastExecutor> castExecutors =
                    fieldTypes.stream()
                            .map(CastExecutors::resolveToString)
                            .collect(Collectors.toList());

            // sorted by partition
            Iterator<InternalRow> iterator =
                    partitions.stream()
                            .map(
                                    partition ->
                                            toRow(
                                                    partition,
                                                    fileStoreTable.partitionKeys(),
                                                    castExecutors,
                                                    fieldGetters,
                                                    fileStoreTable
                                                            .coreOptions()
                                                            .partitionDefaultName()))
                            .sorted(Comparator.comparing(row -> row.getString(0)))
                            .iterator();

            if (readType != null) {
                iterator =
                        Iterators.transform(
                                iterator,
                                row ->
                                        ProjectedRow.from(readType, PartitionsTable.TABLE_TYPE)
                                                .replaceRow(row));
            }
            return new IteratorRecordReader<>(iterator);
        }

        private InternalRow toRow(
                Partition partition,
                List<String> partitionKeys,
                List<CastExecutor> castExecutors,
                InternalRow.FieldGetter[] fieldGetters,
                String defaultPartitionName) {
            PartitionEntry entry = toPartitionEntry(partition);

            StringBuilder partitionStringBuilder = new StringBuilder();

            for (int i = 0; i < partitionKeys.size(); i++) {
                if (i > 0) {
                    partitionStringBuilder.append("/");
                }
                Object partitionValue = fieldGetters[i].getFieldOrNull(entry.partition());
                String partitionValueString =
                        partitionValue == null
                                ? defaultPartitionName
                                : castExecutors
                                        .get(i)
                                        .cast(fieldGetters[i].getFieldOrNull(entry.partition()))
                                        .toString();
                partitionStringBuilder
                        .append(partitionKeys.get(i))
                        .append("=")
                        .append(partitionValueString);
            }

            BinaryString createdByString = BinaryString.fromString(partition.createdBy());
            Timestamp createdAtTimestamp = toTimestamp(partition.createdAt());
            BinaryString updatedByString = BinaryString.fromString(partition.updatedBy());
            BinaryString optionsString = null;
            if (Objects.nonNull(partition.options())) {
                optionsString =
                        BinaryString.fromString(JsonSerdeUtil.toFlatJson(partition.options()));
            }

            return GenericRow.of(
                    BinaryString.fromString(partitionStringBuilder.toString()),
                    partition.recordCount(),
                    partition.fileSizeInBytes(),
                    partition.fileCount(),
                    toTimestamp(partition.lastFileCreationTime()),
                    createdAtTimestamp,
                    createdByString,
                    updatedByString,
                    optionsString);
        }

        private PartitionEntry toPartitionEntry(Partition partition) {
            RowType partitionType = fileStoreTable.schema().logicalPartitionType();
            String defaultPartitionName = fileStoreTable.coreOptions().partitionDefaultName();
            InternalRowSerializer serializer = new InternalRowSerializer(partitionType);
            GenericRow partitionRow =
                    InternalRowPartitionComputer.convertSpecToInternalRow(
                            partition.spec(), partitionType, defaultPartitionName);
            BinaryRow binaryPartition = serializer.toBinaryRow(partitionRow).copy();
            return new PartitionEntry(
                    binaryPartition,
                    partition.recordCount(),
                    partition.fileSizeInBytes(),
                    partition.fileCount(),
                    partition.lastFileCreationTime());
        }

        private Timestamp toTimestamp(Long epochMillis) {
            if (epochMillis == null) {
                return null;
            }
            return Timestamp.fromLocalDateTime(
                    LocalDateTime.ofInstant(
                            Instant.ofEpochMilli(epochMillis), ZoneId.systemDefault()));
        }

        private List<Partition> listPartitions() {
            CatalogLoader catalogLoader = fileStoreTable.catalogEnvironment().catalogLoader();
            if (TimeTravelUtil.hasTimeTravelOptions(new Options(fileStoreTable.options()))
                    || catalogLoader == null) {
                return listPartitionEntries();
            }

            try (Catalog catalog = catalogLoader.load()) {
                Identifier baseIdentifier = fileStoreTable.catalogEnvironment().identifier();
                if (baseIdentifier == null) {
                    return listPartitionEntries();
                }
                String branch = fileStoreTable.coreOptions().branch();
                Identifier identifier;
                if (branch != null && !branch.equals(CoreOptions.BRANCH.defaultValue())) {
                    identifier =
                            new Identifier(
                                    baseIdentifier.getDatabaseName(),
                                    baseIdentifier.getTableName(),
                                    branch);
                } else {
                    identifier = baseIdentifier;
                }
                return catalog.listPartitions(identifier);
            } catch (Exception e) {
                return listPartitionEntries();
            }
        }

        private List<Partition> listPartitionEntries() {
            List<PartitionEntry> partitionEntries =
                    fileStoreTable.newScan().withLevelFilter(level -> true).listPartitionEntries();
            RowType partitionType = fileStoreTable.schema().logicalPartitionType();
            String defaultPartitionName = fileStoreTable.coreOptions().partitionDefaultName();
            String[] partitionColumns = fileStoreTable.partitionKeys().toArray(new String[0]);
            InternalRowPartitionComputer computer =
                    new InternalRowPartitionComputer(
                            defaultPartitionName,
                            partitionType,
                            partitionColumns,
                            fileStoreTable.coreOptions().legacyPartitionName());
            return partitionEntries.stream()
                    .map(entry -> entry.toPartition(computer))
                    .collect(Collectors.toList());
        }
    }
}

File: paimon-python/pypaimon/table/system/partitions_table.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""The ``$partitions`` system table — aggregated partition stats."""

from typing import List, Optional

import pyarrow

from pypaimon.manifest.manifest_file_manager import ManifestFileManager
from pypaimon.manifest.manifest_list_manager import ManifestListManager
from pypaimon.schema.data_types import AtomicType, DataField, RowType
from pypaimon.table.system.system_table import SystemTable


TABLE_TYPE = RowType(False, [
    DataField(0, "partition", AtomicType("STRING", nullable=True)),
    DataField(1, "record_count", AtomicType("BIGINT", nullable=False)),
    DataField(2, "file_size_in_bytes", AtomicType("BIGINT", nullable=False)),
    DataField(3, "file_count", AtomicType("BIGINT", nullable=False)),
    DataField(4, "last_update_time", AtomicType("TIMESTAMP(3)", nullable=True)),
    DataField(5, "created_at", AtomicType("TIMESTAMP(3)", nullable=True)),
    DataField(6, "created_by", AtomicType("STRING", nullable=True)),
    DataField(7, "updated_by", AtomicType("STRING", nullable=True)),
    DataField(8, "options", AtomicType("STRING", nullable=True)),
    DataField(9, "total_buckets", AtomicType("INT", nullable=False)),
    DataField(10, "done", AtomicType("BOOLEAN", nullable=False)),
])


_TIMESTAMP_TYPE = pyarrow.timestamp("ms")


class PartitionsTable(SystemTable):
    """The ``$partitions`` system table.

    The filesystem flow aggregates from manifest entries; catalog-owned
    columns (``created_at``, ``created_by``, ``updated_by``, ``options``,
    ``done``) are filled with placeholders because the filesystem
    catalog does not track them.
    """

    def system_table_name(self) -> str:
        return "partitions"

    def row_type(self) -> RowType:
        return TABLE_TYPE

    def primary_keys(self) -> List[str]:
        return ["partition"]

    def _build_arrow_table(self) -> pyarrow.Table:
        snapshot = self.base_table.snapshot_manager().get_latest_snapshot()
        if snapshot is None:
            return self._empty_table()

        manifest_list_manager = ManifestListManager(self.base_table)
        manifest_files = manifest_list_manager.read_all(snapshot)
        manifest_file_manager = ManifestFileManager(self.base_table)
        entries = manifest_file_manager.read_entries_parallel(
            manifest_files, drop_stats=True)

        partition_map: dict = {}
        for entry in entries:
            spec_items = tuple(
                (field.name, str(value))
                for field, value in zip(
                    entry.partition.fields, entry.partition.values))
            spec_key = spec_items

            stats = partition_map.get(spec_key)
            if stats is None:
                stats = {
                    "spec_items": spec_items,
                    "record_count": 0,
                    "file_size_in_bytes": 0,
                    "file_count": 0,
                    "last_update_time": None,
                    "buckets": set(),
                }
                partition_map[spec_key] = stats

            stats["record_count"] += int(entry.file.row_count)
            stats["file_size_in_bytes"] += int(entry.file.file_size)
            stats["file_count"] += 1
            if entry.file.creation_time is not None:
                ct_ms = entry.file.creation_time.get_millisecond()
                if (stats["last_update_time"] is None
                        or ct_ms > stats["last_update_time"]):
                    stats["last_update_time"] = ct_ms
            stats["buckets"].add(int(entry.bucket))

        partition_strings: List[Optional[str]] = []
        record_counts: List[int] = []
        file_sizes: List[int] = []
        file_counts: List[int] = []
        last_update_times: List[Optional[int]] = []
        total_buckets: List[int] = []

        for stats in partition_map.values():
            partition_strings.append(_render_partition(stats["spec_items"]))
            record_counts.append(stats["record_count"])
            file_sizes.append(stats["file_size_in_bytes"])
            file_counts.append(stats["file_count"])
            last_update_times.append(stats["last_update_time"])
            total_buckets.append(len(stats["buckets"]))

        n = len(partition_map)
        return pyarrow.table({
            "partition": pyarrow.array(
                partition_strings, type=pyarrow.string()),
            "record_count": pyarrow.array(
                record_counts, type=pyarrow.int64()),
            "file_size_in_bytes": pyarrow.array(
                file_sizes, type=pyarrow.int64()),
            "file_count": pyarrow.array(file_counts, type=pyarrow.int64()),
            "last_update_time": pyarrow.array(
                last_update_times, type=_TIMESTAMP_TYPE),
            "created_at": pyarrow.array([None] * n, type=_TIMESTAMP_TYPE),
            "created_by": pyarrow.array([None] * n, type=pyarrow.string()),
            "updated_by": pyarrow.array([None] * n, type=pyarrow.string()),
            "options": pyarrow.array([None] * n, type=pyarrow.string()),
            "total_buckets": pyarrow.array(total_buckets, type=pyarrow.int32()),
            "done": pyarrow.array([False] * n, type=pyarrow.bool_()),
        })

    @staticmethod
    def _empty_table() -> pyarrow.Table:
        return pyarrow.table({
            "partition": pyarrow.array([], type=pyarrow.string()),
            "record_count": pyarrow.array([], type=pyarrow.int64()),
            "file_size_in_bytes": pyarrow.array([], type=pyarrow.int64()),
            "file_count": pyarrow.array([], type=pyarrow.int64()),
            "last_update_time": pyarrow.array([], type=_TIMESTAMP_TYPE),
            "created_at": pyarrow.array([], type=_TIMESTAMP_TYPE),
            "created_by": pyarrow.array([], type=pyarrow.string()),
            "updated_by": pyarrow.array([], type=pyarrow.string()),
            "options": pyarrow.array([], type=pyarrow.string()),
            "total_buckets": pyarrow.array([], type=pyarrow.int32()),
            "done": pyarrow.array([], type=pyarrow.bool_()),
        })


def _render_partition(spec_items) -> Optional[str]:
    """Render a partition spec as ``pt=v/pt2=v2`` or None when empty."""
    if not spec_items:
        return None
    return "/".join("{}={}".format(name, value) for name, value in spec_items)
